Tutorials (#221)
applied-material-modeling/neml2: doc/tutorials/optimization/automatic_differentiation/main.md

@insert-title:tutorials-optimization-automatic-differentiation

[TOC]

## Model definition

> For demonstration purposes, let us go back to our example -- the linear isotropic elasticity model:

Reviewer: ```suggestion
For demonstration purposes, let us go back to an earlier example model -- linear isotropic elasticity:
```

```
[Models]
  [model]
    type = LinearIsotropicElasticity
    strain = 'forces/E'
    stress = 'state/S'
    coefficient_types = 'BULK_MODULUS SHEAR_MODULUS'
    coefficients = '1.4e5 7.8e4'
  []
[]
```

Recall that both the bulk modulus and the shear modulus are treated as *trainable* parameters, which can be verified by inspecting the model summary:
@source:src1
```python
import neml2

model = neml2.load_model("input.i", "model")
print(model)
```
@endsource

Output:
```
@attach-output:src1
```

## Automatic differentiation

NEML2 tensors can be used interchangeably with PyTorch tensors in a seamless fashion. Function graph traced through NEML2 tensor operations can be back-propagated using PyTorch's autograd engine.

For example, the following Python script demonstrates the calculation of \f$\pdv{l}{p}\f$ using an arbitrary, scalar-valued loss function \f$l\f$ obtained by a combination of NEML2 operations and PyTorch operations.

> @source:src2

Reviewer: Strictly speaking it's weird you go to only python at this point and don't have c++, but I understand why given the next section.

Author: Yeah, it's somewhat tricky because we can't use matplotlib in c++, so we'll have to split out the plotting part. After I get to #222 I'll revisit this.
